### PR TITLE
Fix EntityManagerRector based on project_analysis results

### PR DESCRIPTION
**Pull request by bbrala**
## Description
Explain the technical implementation of the work done.

## To Test
- Add steps to test this feature

## Drupal.org issue
Provide a link to the issue from https://www.drupal.org/project/rector/issues. If no issue exists, please create one and link to this PR.

### CODE REVIEW

**Comment by bbrala** [comment]
Fixes 220 error while running analysis.

Search for `vendor/palantirnet/drupal-rector/src/Drupal8/Rector/Deprecation/EntityM`

In https://git.drupalcode.org/project/project_analysis/-/commits/11-55227?ref_type=tags

**Comment by bbrala** [comment]
Marked draft, needs better solution.

**Comment by bbrala** [comment]
Copied a functional test locally, which shows the difference. Will push tonight.

**Comment by bbrala** [comment]
Ok, only error i made was mostly the removal of the depth counter. Also made phpstan a little more happy. Ready for review. Should solve lots of errors in project anlysis

File: src/Drupal8/Rector/Deprecation/EntityManagerRector.php [hunk 1]
```diff
@@ -145,17 +148,18 @@ public function findInstanceByNameInAssign(Node\Expr\Assign $assign, string $cla
         $depth = 0;
 
         // Should the expression be the class we are looking for and the name is the one we are looking for, we can return early.
-        if ($node instanceof $class && $this->getName($node->name) === $name) {
+        if ($node instanceof Node && $node instanceof $class && $this->getName($node->name) === $name) {
```
Comment (mglaman) [inline src/Drupal8/Rector/Deprecation/EntityManagerRector.php:153]: `expr` is nullable? It's not supposed to be

File: src/Drupal8/Rector/Deprecation/EntityManagerRector.php [hunk 2]
```diff
@@ -145,17 +148,18 @@ public function findInstanceByNameInAssign(Node\Expr\Assign $assign, string $cla
...
         }
 
         // Find the relevant class with name in the chain.
-        while (isset($node->var) && !($node->var instanceof $class && $this->getName($node->var->name) !== $name)) {
+        while (isset($node->var) && !($node->var instanceof $class && isset($node->var->name) && $this->getName($node->var->name) !== $name)) {
             $node = $node->var;
             ++$depth;
-            if ($node instanceof $class && $this->getName($node->name)) {
+
+            if ($node instanceof Node && $node instanceof $class && isset($node->name) && $this->getName($node->name)) {
```
Comment (mglaman) [inline src/Drupal8/Rector/Deprecation/EntityManagerRector.php]: again, feels weird that we have to `instanceof` something which shouldn't be nullable and is `\PhpParser\Node\Expr`

And we can call `$this->getName($node)`, that's what the name resolver is for.

File: src/Drupal8/Rector/Deprecation/EntityManagerRector.php [hunk 3]
```diff
@@ -145,17 +148,18 @@ public function findInstanceByNameInAssign(Node\Expr\Assign $assign, string $cla
...
-        if ($node instanceof $class && $this->getName($node->name) === $name) {
+        if ($node instanceof Node && $node instanceof $class && $this->getName($node->name) === $name) {
             $node->setAttribute(self::class, $depth);
 
             return $node;
         }
 
         // Find the relevant class with name in the chain.
-        while (isset($node->var) && !($node->var instanceof $class && $this->getName($node->var->name) !== $name)) {
+        while (isset($node->var) && !($node->var instanceof $class && isset($node->var->name) && $this->getName($node->var->name) !== $name)) {
```
Comment (mglaman) [inline src/Drupal8/Rector/Deprecation/EntityManagerRector.php:160]: Would this work if we did `$this->getName($node->var)`? See `\Rector\NodeNameResolver\NodeNameResolver\VariableNameResolver::resolve`

File: tests/src/Drupal8/Rector/Deprecation/EntityManagerRector/EntityManagerRectorTest.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+     * @return Iterator<<string>>
+     */
+    public static function provideData(): \Iterator
+    {
+        return self::yieldFilesFromDirectory(__DIR__.'/fixture');
+    }
+
+    public function provideConfigFilePath(): string
+    {
+        // must be implemented
```
Comment (mglaman) [inline tests/src/Drupal8/Rector/Deprecation/EntityManagerRector/EntityManagerRectorTest.php]: it has?

File: tests/src/Drupal8/Rector/Deprecation/EntityManagerRector/config/configured_rule.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+declare(strict_types=1);
+
+use DrupalRector\Drupal8\Rector\Deprecation\EntityManagerRector;
+use DrupalRector\Drupal8\Rector\Deprecation\EntityViewRector;
+use DrupalRector\Tests\Rector\Deprecation\DeprecationBase;
+use Rector\Config\RectorConfig;
+
+return static function (RectorConfig $rectorConfig): void {
+    // $rectorConfig->rule(EntityViewRector::class);
```
Comment (mglaman) [inline tests/src/Drupal8/Rector/Deprecation/EntityManagerRector/config/configured_rule.php]: we can remove this?

**Review by mglaman — CHANGES_REQUESTED** [review]
Notes, I think we can tidy up some of the checks.

File: src/Drupal8/Rector/Deprecation/EntityManagerRector.php (hunk 3 above)
Comment (bbrala) [reply src/Drupal8/Rector/Deprecation/EntityManagerRector.php:160]: I tried this, it makes the functional tests fail :(

**Comment by bbrala** [comment]
I tried to use getName properly, but couldnt get unit and functional tests to work :( 

(small tip: `brew install act`, then remove the unneeded matrix definitions and run: `act push -W ./.github/workflows/functional_test__rector_examples.yml`, functional test locally <3)

**Review by mglaman — APPROVED** [review]
Approving, too much work to make it fancy :)

**Pull request merged**